OTA-802: Refactor channel guidance (#47333)
openshift/openshift-docs: modules/understanding-upgrade-channels.adoc

[id="conditional-updates-overview_{context}"]
== Update recommendation removals and Condtional Updates
Red Hat monitors newly released versions and update paths associated with those versions both before and after they're added to supported channels. If a serious regression is identified Red Hat may remove affected update recommendations. When Red Hat chooses to remove update recommendations that action is taken in all relevant channels, simultaneously. The removal of recommended updates may happen either before or after updates have been promoted to supported channels.

In the event that Red Hat removes update recommendations from any supported release we will eventually provide a superseding update recommendation to a future version which corrects the regression. There may however be a delay while the defect is corrected, tested, and promoted to your selected channel.

> Beginning with {product-title} 4.10 whenever update recommendations are removed from supported channels they are replaced with Conditional Updates which declare one or more known risks. Each known risk may apply to all clusters or only clusters matching certain conditions such as having `Platform` set to `None`` or the CNI provider set to `OpenShiftSDN`. The Cluster Version Operator (CVO) continually evaluates known risks against the current cluster state. When no risks match, the update is recommended. You can update from the Administrator perspective on the web console, as well as the OpenShift CLI (`oc`).

Reviewer: nit: doubled backtick after `None`

> When an update is not recommended because a risk might apply, you can view the update from the OpenShift CLI (`oc`). If the cluster administrator evaluates the potential known risks and decides it is acceptable for the current cluster, the administrator can waive the safety guards and proceed to an update.

Reviewer: The pre-existing `oc` focus here was because we landed this language before the console grew a UX for updating despite matching risks.  They [backported that UX in 4.10.16][1], and these docs are aimed at 4.12 and later, so we can probably either:

* Mention "the Administrator perspective on the web console" in this paragraph, or
* Drop all client references from both this and the previous paragraph, focusing on the workflow semantics, and leaving it implicit that both of our update clients support all of the discussed workflows.

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=2087041#c5

Author: PTAL, I've done the bare minimum but we should probably take a card to work with docs to walk through the web console workflow, but apparently we're not allowed to use screenshots so that'll be fun.

Second, you should choose your desired rollout strategy. You may choose to update as soon as Red Hat declares a release GA by selecting from fast channels or you may wish to wait for Red Hat to promote releases to the stable channel. Update recommendations offered in the `fast-{product-version}` and `stable-{product-version}` are both fully supported and benefit equally from ongoing data analysis. The promotion delay before promoting release to the stable channel represets the only difference between the two channels. Updates to the latest z-streams are generally promoted to the stable channel within a week or two, however the delay when initially rolling out updates to the latest minor is much longer, generally 45-90 days. Please consider the promotion delay when choosing your desired channel as waiting for promotion to the stable channel may affect your scheduling plans.

> Additionally, there are several factors which may lead an organization to move clusters to the fast channel either permanently or temporarilly including:

Reviewer: "temporarilly" -> ["temporarily"][1]?

[1]: https://en.wiktionary.org/wiki/temporarily

In addition to the stable channel, all even-numbered minor versions of {product-title} offer link:https://access.redhat.com/support/policy/updates/openshift#ocp4_phases[Extended Update Support] (EUS). Releases which are promoted to the stable channel are also simultaneously promoted to the EUS channels. The primary purpose of the EUS channels is to serve as a convenience for clusters performing an EUS-to-EUS update.

[NOTE]
====
Both standard and non-EUS subscribers can access all EUS repositories and necessary RPMs (`rhel-*-eus-rpms`) to be able to support critical purposes such as debugging and building drivers.
====

[id="candidate-version-channel_{context}"]
== candidate-{product-version} channel

The `candidate-{product-version}` channel offers unsupported early access to releases as soon as they're built. Releases present only in candidate channels
may not contain the full feature set of eventual GA releases or features may be removed prior to GA. Additionally, these releases have not been subject to full
Red Hat Quality Assurance and may not offer update paths to later GA releases. Given these caveats the candidate channel is only suitable for testing purposes

> where destroying and recreating a cluster is acceptable.

Author: @julienlim here's the updated section on candidate channels
- move unsupported'ness front and center
- drops the assertion that items in candidate are always feature complete and grants us ability to drop features

[id="understanding-upgrade-channels_{context}"]

= Update channels

ifndef::openshift-origin[]
[id="fast-version-channel_{context}"]
== fast-{product-version} channel

> The `fast-{product-version}` channel is updated with new versions of {product-title} {product-version} as soon as Red Hat declares the version as a general availability (GA) release. As such, these releases are fully commercially supported and purposed to be used in production environments.

Reviewer 2: I added some details in the `supported` vs `commercially supported` thread. `Just `supported` would be the best practice here

Author: Ok, I'll give Tim time to respond and then move forward with your suggestion as that's what I had originally but product experience requested the formal terminology.

[id="stable-version-channel_{context}"]
== stable-{product-version} channel
While the `fast-{product-version}` channel contains releases as soon as their errata are published, releases are added to the `stable-{product-version}` channel after a delay. During this delay, data is collected from multiple sources and analyzed for indications of product regressions. Once a significant number of data points have been collected, and absent negative signals, these releases are added to the stable channel.

[NOTE]

> Since the time required to obtain a significant number of data points varies based on many factors, Service LeveL Objective (SLO) is not offered for the delay duration between fast and stable channels. Please see xref:../updating/understanding-upgrade-channels-release.html#fast-stable-channel-strategies_{context}[Choosing the correct channel for your cluster] for more information.

Reviewer 2: Also made a comment about this xref in the original comment

where destroying and recreating a cluster is acceptable.

ifdef::openshift-origin[]
[id="stable-4-channel_{context}"]
== stable-4 channel
Releases are added to the `stable-4` channel after passing all tests and stable-4 is the only supported channel.
endif::openshift-origin[]

[id="upgrade-version-paths_{context}"]
== Update recommendations in the channel

{product-title} maintains an update recommendation service that knows your installed {product-title} version and the path to take within the channel to get you to the next release. Update paths are also limited to versions relevant to your currently selected channel and its promotion characteristics.

You can imagine seeing the following releases in your channel:

* {product-version}.0
* {product-version}.1
* {product-version}.3
* {product-version}.4

The service recommends only updates that have been tested and have no known serious regressions. For example, if your cluster is on {product-version}.1 and {product-title} suggests {product-version}.4, then it is recommended to update from {product-version}.1 to {product-version}.4. 

[IMPORTANT]
====

> Do not rely on consecutive patch numbers. In this example, {product-version}.2 is not and never was available in the channel, therefore updates to {product-version}.2 are neither recommended nor supported.

Reviewer 2: ```suggestion
Do not rely on consecutive patch numbers. In this example, {product-version}.2 is not and never was available in the channel, therefore updates to {product-version}.2 are not recommended or supported.
```

If Red Hat removes update recommendations from any supported release, a superseding update recommendation will be provided to a future version that corrects the regression. There may however be a delay while the defect is corrected, tested, and promoted to your selected channel.

Beginning in {product-title} 4.10, when update recommendations are removed from supported channels, they are replaced with Conditional Updates that declare one or more known risks. Each known risk may apply to all clusters or only clusters matching certain conditions. Some examples include having the `Platform` set to `None` or the CNI provider set to `OpenShiftSDN`. The Cluster Version Operator (CVO) continually evaluates known risks against the current cluster state. If no risks match, the update is recommended. If the risk matches, those updates are listed as a `Supported But Not Recommended` update and a reference link is provided. The reference link helps the cluster admin decide if they would like to accept the risk and update anyway.

[role="_additional-resources"]

> .Additional resources

Reviewer 2: The additional resources sections go in the assembly file, here the file would be: updating/understanding-upgrade-channels-release.adoc. Becuase we can't put `xrefs` in modules

Author: I've moved it to the assembly.

[NOTE]
====
Due to the complexity involved in planning upgrades between versions many minors apart, channels that assist in planning upgrades beyond a single EUS-to-EUS update are not offered.
====

Second, you should choose your desired rollout strategy. You may choose to update as soon as Red Hat declares a release GA by selecting from fast channels or you may want to wait for Red Hat to promote releases to the stable channel. Update recommendations offered in the `fast-{product-version}` and `stable-{product-version}` are both fully supported and benefit equally from ongoing data analysis. The promotion delay before promoting release to the stable channel represets the only difference between the two channels. Updates to the latest z-streams are generally promoted to the stable channel within a week or two, however the delay when initially rolling out updates to the latest minor is much longer, generally 45-90 days. Please consider the promotion delay when choosing your desired channel as waiting for promotion to the stable channel may affect your scheduling plans.

> Additionally, there are several factors which may lead an organization to move clusters to the fast channel either permanently or temporarily including:

Reviewer 2: I believe this is meant to be formatted as bullet points. If so, you need a space between lines 101 and 102 for it to render as a unordered list

 - The desire to apply a specific fix known to affect your environment without delay.

>  - Application of CVE fixes without delay -- even CVE fixes may introduce regressions, so we still apply promotion delays to z-streams with CVE fixes.

Reviewer 2: ```suggestion
 - Application of CVE fixes without delay. CVE fixes may introduce regressions, so promotion delays still apply to z-streams with CVE fixes.
```

>  - Internal testing processes -- If it takes your organization several weeks to qualify releases it's best test concurrently with our promotion process rather than waiting. This also assures that any telemetry signal provided to Red Hat is a factored into our rollout, we can fix issues relevant to you faster.

Reviewer 2: ```suggestion
 - Internal testing processes. If it takes your organization several weeks to qualify releases it is best test concurrently with our promotion process rather than waiting. This also assures that any telemetry signal provided to Red Hat is a factored into our rollout, so issues relevant to you can be fixed faster.
```

[id="understanding-upgrade-channels_{context}"]

= Update channels

> [id="candidate-version-channel_{context}"]

Reviewer: you'll want to move this anchor to follow your shifting of the `candidate-4.y` section.

= Update channels

[id="candidate-version-channel_{context}"]
== candidate-{product-version} channel

> The `candidate-{product-version}` channel contains candidate builds for a z-stream ({product-version}.z) and previous minor version releases. Release candidates contain all the features of the product but are not supported. Use release candidate versions to test feature acceptance and assist in qualifying the next version of {product-title}. A release candidate is any build that is available in the candidate channel, including ones that do not contain link:https://semver.org/spec/v2.0.0.html#spec-item-9[a pre-release version] such as `-rc` in their names. After a version is available in the candidate channel, it goes through more quality checks. If it meets the quality standard, it is promoted to the `fast-{product-version}` or `stable-{product-version}` channels. Because of this strategy, if a specific release is available in both the `candidate-{product-version}` channel and in the `fast-{product-version}` or `stable-{product-version}` channels, it is a Red Hat-supported version. The `candidate-{product-version}` channel can include release versions from which there are no recommended updates in any channel.

Reviewer 3: Original phrasing too wordy (and puts the important bits at the end, i.e. not supported):
Release candidates contain all the features of the product but are not supported. 

Suggestion:
Release candidates are not supported, but they contain all product features.

Reviewer: we are also trying to move away from "but they contain all product features" for prereleases, as we approach 4.13 and possibly putting 4.13.0-ec.# into a `candidate-4.13` channel.  Those engineering candidates will still be in the prerelease channel, but because we'll be cutting them basically as soon as 4.12 forks off, they won't contain all the features that will eventually be part of 4.13.

Author: I'll try to clean it up a bit below and tag you for comment there.

[NOTE]
====
Both standard and non-EUS subscribers can access all EUS repositories and necessary RPMs (`rhel-*-eus-rpms`) to be able to support critical purposes such as debugging and building drivers.
====

== candidate-{product-version} channel

The `candidate-{product-version}` channel contains releases as soon as they're produced, prior to being declared GA. Releases present only in the candidate channels have not completed full Quality Assurance testing and lack Red Hat Support. Therefore releases present only in the candidate channels are only suitable for testing purposes where destroying and recreating a cluster is acceptable. Additionally, we provide no guarantee that these releases will be offered supported update paths to GA versions.

ifdef::openshift-origin[]
[id="stable-4-channel_{context}"]
== stable-4 channel
Releases are added to the `stable-4` channel after passing all tests and stable-4 is the only supported channel.
endif::openshift-origin[]

[id="upgrade-version-paths_{context}"]

> == Update version paths

Author: I'd sort of like to get rid of it because it's pretty graph centric and if we're trying to just re-orient toward update recommendations maybe it's no longer necessary.

@wking @LalatenduMohanty thoughts?

Reviewer: A lot of the other sections are more release-focused (nodes), and this section is update-focused (edges).  I think there's some utility to have a section focused on positive update recommendations to remind everyone of why we're here before moving on to talk about recommendation-removal and update risks.

I agree that the current approach within the section is graph-y.  Perhaps we could rephrase to be more specifically about a single hypothetical cluster running `{product-version}.0` in our hypothetical channel content?

Reviewer 4: The section title `Update version paths` does not seem right here. We should rename as `Upgrade recommendations in the channel`. The content seems relevant to me.

{product-title} maintains an update recommendation service that understands the version of {product-title} you have installed as well as the path to take within the channel you choose to get you to the next release. Update paths are also limited to versions relevant to your currently selected channel and its promotion characteristics.

You can imagine seeing the following releases in your channel:

* {product-version}.0
* {product-version}.1
* {product-version}.3
* {product-version}.4

The service recommends only updates that have been tested and have no known serious regressions. For example, if your cluster is on {product-version}.1 and {product-title} suggests {product-version}.4, then it is recommended to update from {product-version}.1 to {product-version}.4. Do not rely on consecutive patch numbers. In this example, {product-version}.2 is not and never was available in the channel, therefore updates to {product-version}.2 are neither recommended nor supported.

[id="update-recommendation-removal"]
== Update recommendation removals

> Red Hat monitors newly released versions and update paths associated with those versions both before and after they're added to supported channels. When a serious regression is identified Red Hat may remove affected update recommendations. If Red Hat chooses to remove update recommendations that action is taken in all relevant channels, simultaneously. The removal of recommended updates may happen either before or well after updates have been offered in supported channels.

Reviewer 4: I think just after is understood. So I do not see a need to do changes here

> In the event that Red Hat removes update recommendations from any supported release we will eventually provide a superseding update recommendation to a future version which corrects the regression. There may however be a delay while the defect is corrected, tested, and promoted.

Reviewer: nit: possibly "promoted to supported channels"?  Or "your channel"?  Or something that gives more context for "promoted" for folks who may not live in a "for update channels" mindset ;), without getting too far into the weeds about which releases will eventually end up in which channels?

When update recommendations are removed from supported channels those update paths remain supported but are no longer recommended. All supported but not recommended update paths updating from {product-title} version 4.10 and later will have a descriptive message and reference URL which links to more thorough explanation for the recommendation removal. Please review the linked reference and carefully consider whether you wish to continue with an update which is no longer recommended. See link:conditional-updates-overview_{context}[Conditional Updates] section for more information on this feature.

ifndef::openshift-origin[]

[id="fast-stable-channel-strategies_{context}"]
== Choosing the correct channel for your cluster

> Because update recommendations available in both `fast-{product-version}` and `stable-{product-version}` are fully supported and benefit equally from extensive data analysis the choice between the two channels should be guided based on your specific needs rather than which channel is the default or qualities implied by the channel names. Largely this is a choice between updating to the latest GA versions as soon as Red Hat is willing to offer support or waiting while Red Hat collects additional data before promoting to the stable channel.

Reviewer: One way to break up the first sentence here would be:

> Update recommendations available in both `fast-{product-version}` and `stable-{product-version}` are fully supported and benefit equally from extensive data analysis.  The choice between...

And although both default-inertia and name-emotions are part of some channel-selection discussion, I'm a bit squeamish about calling that out directly.  Maybe skip over those, and get right into delay?  And possibly touch on regression risk (present in both, but slightly higher in fast) vs. latency (slightly to significantly higher in stable)?

Author: I took a stab at combining the rollout delay and channel selection into one section that covers them together. PTAL, happy to revert if this seems more confusing.

Additionally, there are several factors which may lead an organization to move clusters to the fast channel either permanently or temporarilly including:
 - The desire to apply a specific fix known to affect your environment without delay.

>  - Application of CVE fixes without delay -- Because even CVE fixes may introduce regressions we still apply promotion delays to z-streams with CVE fixes.

Reviewer: possible rephrasing:

> even CVE fixes may introduce regressions, so we still apply promotion delays to z-streams with CVE fixes.

 - Internal pre-prod promotion processes -- If it takes your organization several weeks to promote updates between environments it would be best to use the latest available rather than introducing additional weeks or months of delay.

=== Minor version initial update rollouts

> Because the scope of changes associated with a z-stream update is orders of magnitude lower than the scope of changes between any two minor versions our data collection requirements are similarly orders of magnitude different as well. While we can safely promote z-stream updates within a week or two of GA we cannot do the same with minor version updates. Therefore, it's expected that the initial availability of update recommendations to the latest minor may take 45-90 days to appear in the stable channel.

Reviewer: nit: we talk about z-stream and minor a lot in this doc, so probably fine how you have it.  But possibly worth providing examples occasionally like "z-stream updates (within a 4.y)" and "between any two minor versions (4.y to 4.(y+1))" and similar, for folks who haven't internalized [SemVer's major.minor.patch naming][1].

[1]: https://semver.org/spec/v2.0.0.html#summary

> Given this lengthy data collection phase it's important that any organization which has their own lengthy production qualification processes begin testing at GA rather than waiting for stable channel update recommendations to become available. Waiting for stable channel update edges only further delays defect resolution of any defect unique to your usecase or environment.

Reviewer 4: We should mention fast channel if we are mentioning stable channel.  Suggested text `any organization which has their own lengthy production qualification processes begin testing using fast channel at GA rather than waiting for stable channel`

Author: I got rid of this paragraph in a refactoring and I think this paragraph covers your suggestion now. https://github.com/openshift/openshift-docs/pull/47333/files#diff-a730a4f4ac2c41dde208df2277ff16a4411c015a889dc4ba15323ed8805df5ceR91-R94

Given this lengthy data collection phase it's important that any organization which has their own lengthy production qualification processes begin testing at GA rather than waiting for stable channel update recommendations to become available. Waiting for stable channel update edges only further delays defect resolution of any defect unique to your usecase or environment.

The `fast-{product-version}` and `stable-{product-version}` channels present a choice between receiving general availability releases as soon as they are available or allowing Red Hat to control the rollout of those updates. If issues are detected during rollout or at a later time, upgrades to that version might be blocked in both the `fast-{product-version}` and `stable-{product-version}` channels, and a new version might be introduced that becomes the new preferred upgrade target.

> Customers can improve this process by configuring pre-production systems on the `fast-{product-version}` channel, configuring production systems on the `stable-{product-version}` channel, and participating in the Red Hat connected customer program. Red Hat uses this program to observe the impact of updates on your specific hardware and software configurations. Future releases might improve or alter the pace at which updates move from the `fast-{product-version}` to the `stable-{product-version}` channel.

Reviewer: Pitching [remote health monitoring][1] (which is what I think "participating in the Red Hat connected customer program" was trying to do?) seems useful, and I don't see it mentioned in your proposed text yet.  But :shrug: I'm fine if folks want to drop that pitch too.

[1]: https://docs.openshift.com/container-platform/4.11/support/remote_health_monitoring/about-remote-health-monitoring.html

[id="understanding-upgrade-channels_{context}"]

= Update channels

ifndef::openshift-origin[]
[id="fast-version-channel_{context}"]
== fast-{product-version} channel
The `fast-{product-version}` channel is updated with new versions of {product-version} as soon as Red Hat declares the given version as a general availability (GA) release. As such, these releases are fully commercially supported and purposed to be used in production environments.

[id="stable-version-channel_{context}"]
== stable-{product-version} channel
While the `fast-{product-version}` channel contains releases as soon as their errata are published, releases are added to the `stable-{product-version}` channel after a delay. During this delay, data is collected from multiple sources and continuously analyzed for indications of product regressions. Once a statistically significant number of data points have been collected and absent negative signals these releases are added to the stable channel.

[NOTE]

> Because the time required to obtain a statistically significant number of data points varies based on many factors we offer no Service LeveL Objective (SLO) for the delay duration between fast and stable channels. Please see link:strategies_{context}[Fast and stable channel use and strategies] for more information.

Author: I used xref to an anchor as it's within the same module, does that work?

> Because the time required to obtain a statistically significant number of data points varies based on many factors we offer no Service LeveL Objective (SLO) for the delay duration between fast and stable channels. Please see link:strategies_{context}[Fast and stable channel use and strategies] for more information.

Reviewer 2: So we can't use xrefs in module files because it can mess with the software we use to deploy the website. In this case you can say. 

`For more information, please see "Choosing the correct channel for your cluster"`

Then add the xref to the additional resources section in the assembly file (updating/understanding-upgrade-channels-release.adoc)

may not contain the full feature set of eventual GA releases or features may be removed prior to GA. Additionally, these releases have not been subject to full
Red Hat Quality Assurance and may not offer update paths to later GA releases. Given these caveats the candidate channel is only suitable for testing purposes
where destroying and recreating a cluster is acceptable.

ifdef::openshift-origin[]
[id="stable-4-channel_{context}"]
== stable-4 channel
Releases are added to the `stable-4` channel after passing all tests and stable-4 is the only supported channel.
endif::openshift-origin[]

[id="upgrade-version-paths_{context}"]
== Update version paths

{product-title} maintains an update recommendation service that understands the version of {product-title} you have installed as well as the path to take within the channel you choose to get you to the next release. Update paths are also limited to versions relevant to your currently selected channel and its promotion characteristics.

You can imagine seeing the following releases in your channel:

* {product-version}.0
* {product-version}.1
* {product-version}.3
* {product-version}.4

> The service recommends only updates that have been tested and have no known serious regressions. For example, if your cluster is on {product-version}.1 and {product-title} suggests {product-version}.4, then it is recommended to update from {product-version}.1 to {product-version}.4. Do not rely on consecutive patch numbers. In this example, {product-version}.2 is not and never was available in the channel, therefore updates to {product-version}.2 are neither recommended nor supported.

Reviewer 2: The `Do not rely on consecutive patch numbers` might be better as a [IMPORTANT] block afterwards, but thats up to you!

[id="conditional-updates-overview_{context}"]
== Update recommendation removals and Condtional Updates
Red Hat monitors newly released versions and update paths associated with those versions both before and after they're added to supported channels. If a serious regression is identified Red Hat may remove affected update recommendations. When Red Hat chooses to remove update recommendations that action is taken in all relevant channels, simultaneously. The removal of recommended updates may happen either before or after updates have been promoted to supported channels.

In the event that Red Hat removes update recommendations from any supported release we will eventually provide a superseding update recommendation to a future version which corrects the regression. There may however be a delay while the defect is corrected, tested, and promoted to your selected channel.

> Beginning with {product-title} 4.10 whenever update recommendations are removed from supported channels they are replaced with Conditional Updates which declare one or more known risks. Each known risk may apply to all clusters or only clusters matching certain conditions such as having `Platform` set to `None` or the CNI provider set to `OpenShiftSDN`. The Cluster Version Operator (CVO) continually evaluates known risks against the current cluster state. When no risks match, the update is recommended. If the risk matches those updates are listed as Supported But Not Recommended Updates, a reference link is provided which helps the cluster admin decide if they would like to accept the risk and update anyway.

Reviewer: nit: ", and a reference link is provided..." (adding `and`)?

> Beginning with {product-title} 4.10 whenever update recommendations are removed from supported channels they are replaced with Conditional Updates which declare one or more known risks. Each known risk may apply to all clusters or only clusters matching certain conditions such as having `Platform` set to `None` or the CNI provider set to `OpenShiftSDN`. The Cluster Version Operator (CVO) continually evaluates known risks against the current cluster state. When no risks match, the update is recommended. If the risk matches those updates are listed as Supported But Not Recommended Updates and a reference link is provided which helps the cluster admin decide if they would like to accept the risk and update anyway.

Author: Aside from calling Platform an object and `None` a status I'm taking this suggestion as is. Platform is a field on the infrastructure object and I'm being a bit loose with context here so as to avoid being overly wordy, hope that's ok.

[role="_additional-resources"]
.Additional resources
* For more information, see xref:../updating/updating-cluster-cli.adoc#update-conditional-upgrade-pathupdating-cluster-cli[Updating along a conditional upgrade path]

ifndef::openshift-origin[]

[id="fast-stable-channel-strategies_{context}"]

> == Choosing the correct channel for your cluster

Reviewer 2: This may need to be broken up into a procedure module

Choosing the appropriate channel involves two decisions.

> First, select the minor version you wish for your cluster to upgrade to. Selecting a channel which matches your current version ensures that you only apply z-stream updates and do not receive feature updates. Selecting an available channel which has a version greater than your current version will ensure that after one or more updates your cluster will have updated to that version. Your cluster will only be offered channels which match its current version, the next version, or the next EUS version. Due to the complexity involved in planning upgrades between versions many minors apart we do not offer channels which assist in planning upgrades beyond a single EUS-to-EUS update.

Reviewer 2: `wish` -> `want`

Change
`Due to the complexity involved in planning upgrades between versions many minors apart we do not offer channels which assist in planning upgrades beyond a single EUS-to-EUS update` to 

`Due to the complexity involved in planning upgrades between versions many minors apart, channels that assist in planning upgrades beyond a single EUS-to-EUS update are not offered.`

Author: Ok, what do you think about having the `Due to the complexity...` bit as an admonition?

Author: I went ahead with a NOTE admonition.

> Second, you should choose your desired rollout strategy. You may choose to update as soon as Red Hat declares a release GA by selecting from fast channels or you may wish to wait for Red Hat to promote releases to the stable channel. Update recommendations offered in the `fast-{product-version}` and `stable-{product-version}` are both fully supported and benefit equally from ongoing data analysis. The promotion delay before promoting release to the stable channel represets the only difference between the two channels. Updates to the latest z-streams are generally promoted to the stable channel within a week or two, however the delay when initially rolling out updates to the latest minor is much longer, generally 45-90 days. Please consider the promotion delay when choosing your desired channel as waiting for promotion to the stable channel may affect your scheduling plans.

Reviewer 2: `wish` -> `want`

ifndef::openshift-origin[]

[id="fast-stable-channel-strategies_{context}"]
== Choosing the correct channel for your cluster

Because update recommendations available in both `fast-{product-version}` and `stable-{product-version}` are fully supported and benefit equally from extensive data analysis the choice between the two channels should be guided based on your specific needs rather than which channel is the default or qualities implied by the channel names. Largely this is a choice between updating to the latest GA versions or waiting while Red Hat collects additional data before promoting to the stable channel.

> Customers can improve this process by configuring pre-production systems on the `fast-{product-version}` channel, configuring production systems on the `stable-{product-version}` channel, and participating in the Red Hat connected customer program. Red Hat uses this program to observe the impact of updates on your specific hardware and software configurations.

Author: Need to clean this paragraph up as it implies that fast is only suitable for pre-production.

[id="understanding-upgrade-channels_{context}"]

= Update channels

[id="candidate-version-channel_{context}"]

ifndef::openshift-origin[]
[id="fast-version-channel_{context}"]
== fast-{product-version} channel

> The `fast-{product-version}` channel is updated with new versions of {product-version} as soon as Red Hat declares the given version as a general availability (GA) release. As such, these releases are fully supported, are production quality, and have had errata notices delivered.

Reviewer 5: "As such, these releases are fully commercially supported and purposed to be used in production environments."

> The `fast-{product-version}` channel is updated with new versions of {product-version} as soon as Red Hat declares the given version as a general availability (GA) release. As such, these releases are fully commercially supported and purposed to be used in production environments.

Author: @reestr made this suggestion, will let him speak to that.

> The `fast-{product-version}` channel is updated with new versions of {product-version} as soon as Red Hat declares the given version as a general availability (GA) release. As such, these releases are fully commercially supported and purposed to be used in production environments.

Reviewer 4: I still think we should say `supported` instead of `commercially supported`

> The `fast-{product-version}` channel is updated with new versions of {product-version} as soon as Red Hat declares the given version as a general availability (GA) release. As such, these releases are fully commercially supported and purposed to be used in production environments.

Author: @reestr Can you provide input here as to how important you feel your `commercially supported` word choice is?

> The `fast-{product-version}` channel is updated with new versions of {product-version} as soon as Red Hat declares the given version as a general availability (GA) release. As such, these releases are fully commercially supported and purposed to be used in production environments.

Reviewer 5: @sdodson - lets keep supported then to ensure documentation consistency.

[id="stable-version-channel_{context}"]
== stable-{product-version} channel
While the `fast-{product-version}` channel contains releases as soon as their errata are published, releases are added to the `stable-{product-version}` channel after a delay. During this delay, data is collected from multiple sources and continuously analyzed for indications of product regressions. Once a statistically significant number of data points have been collected and absent negative signals these releases are added to the stable channel.

[NOTE]
Because the time required to obtain a statistically significant number of data points varies based on many factors we offer no SLA for the delay duration between fast and stable channels. Please see link:strategies_{context}[Fast and stable channel use and strategies] for more information.

Newly installed clusters default to utilizing stable channels.

[id="eus-4y-channel_{context}"]
== eus-4.y channel

> In addition to the stable channel, all even-numbered minor versions of {product-title} offer an link:https://access.redhat.com/support/policy/updates/openshift#ocp4_phases[Extended Update Support] (EUS). Releases which are promoted to the stable channel are also simultaneously promoted to the EUS channels. The primary purpose of the EUS channels is to serve as a convenience for clusters performing an EUS-to-EUS update.

Reviewer 6: Remove `an`:
> offer Extended Update Support (EUS).

> In addition to the stable channel, all even-numbered minor versions of {product-title} offer an link:https://access.redhat.com/support/policy/updates/openshift#ocp4_phases[Extended Update Support] (EUS). Releases which are promoted to the stable channel are also simultaneously promoted to the EUS channels. The primary purpose of the EUS channels is to serve as a convenience for clusters performing an EUS-to-EUS update.

Author: If you stay on the eus-4.8 channel you will never see 4.9 versions and we don't have an eus-4.9 channel. The convenience is that you switch from eus-4.8 to eus-4.10 and click the upgrade button twice you'll end up on 4.10.

After you have upgraded from 4.8 to 4.9 but before upgrading to 4.10 you could change your mind and switch to a 4.9 channel.

[NOTE]
====
Both standard and non-EUS subscribers can access all EUS repositories and necessary RPMs (`rhel-*-eus-rpms`) to be able to support critical purposes such as debugging and building drivers.
====

== candidate-{product-version} channel

> The `candidate-{product-version}` channel contains releases as soon as they're produced, prior to being declared GA. Releases present only in the candidate channels have not completed full Quality Assurance testing and lack Red Hat Support. Therefore releases present only in the candidate channels are only suitable for testing purposes where destroying and recreating a cluster is acceptable. Additionally, we provide no guarantee that these releases will be offered update paths to GA versions.

Reviewer: Can we avoid talking about unqualified "update paths" and instead focus on "recommended update paths" and "supported update paths" (with recommended being a subset of supported paths).  Or maybe "update paths in the `candidate-{product-version}` channel"?  Or something that makes it more clear that folks on candidate releases we think are hopeless can still ask their cluster to update to other versions, regardless of whats in any channel, and their cluster will give it a shot.

ifdef::openshift-origin[]
[id="stable-4-channel_{context}"]
== stable-4 channel
Releases are added to the `stable-4` channel after passing all tests and stable-4 is the only supported channel.
endif::openshift-origin[]

[id="upgrade-version-paths_{context}"]
== Update version paths

{product-title} maintains an update recommendation service that understands the version of {product-title} you have installed as well as the path to take within the channel you choose to get you to the next release. Update paths are also limited to versions relevant to your currently selected channel and its promotion characteristics.

You can imagine seeing the following releases in your channel:

* {product-version}.0
* {product-version}.1
* {product-version}.3
* {product-version}.4

> The service recommends only updates that have been tested and have no known serious regressions. For example, if your cluster is on {product-version}.1 and {product-title} suggests {product-version}.4, then it is safe for you to update from {product-version}.1 to {product-version}.4. Do not rely on consecutive patch numbers. In this example, {product-version}.2 is not and never was available in the channel, therefore updates to {product-version}.2 are neither recommended nor supported.

Reviewer 4: `{product-version}.2 is not and never was available in the channel, therefore updates to {product-version}.2 are neither recommended nor supported.` has some caveat. We should say ` if {product-version}.2 is not and never was released and not available in the channel, therefore updates to {product-version}.2 are neither recommended nor supported.`

Reviewer: this section covers `candidate-*` too, so "then it is safe" -> "then it is recommended"?  Or something else that leaves more room for "given the stability commitments of your configured channel".

Author: @wking ok, will make that change. I'd also considered whether or not we need to get into this much detail here. If the cluster is connected and offering updates then they're recommended, in this context they'll never know that "{product=version}.2" existed and I'm not too worried about the risk of them hammering that screw in.

[id="update-recommendation-removal"]
== Update recommendation removals
Red Hat monitors newly released versions and update paths associated with those versions both before and after they're added to supported channels. When a serious regression is identified Red Hat may remove affected update recommendations. If Red Hat chooses to remove update recommendations that action is taken in all relevant channels, simultaneously. The removal of recommended updates may happen either before or well after updates have been offered in supported channels.

> In the event that Red Hat removes update recommendations from any supported release we will eventually provide a superseding update paths to a future version which corrects the regression. There may however be a delay while the defect is corrected, tested, and promoted.

Reviewer 6: > superseding update **_path_**

> When update recommendations are removed from supported channels those update paths remain supported but are no longer recommended. All supported but not recommended update paths updating from {product-title} version 4.10 and later will have a descriptive message and reference URL which links to more thorough explanation for the recommendation removal. Please review the linked reference and carefully consider whether you wish to continue with an update which is no longer recommended. See link:conditional-updates-overview_{context}[Conditional Updates] section for more information on this feature.

Reviewer 4: It is not clear that we are talking about conditional updates in the beginning. We should say that in 4.10 or later versions we may conditionally recommend update paths. As part of the conditional update , the condition is evaluated and then CVO decided if the issue is applicable to the cluster or not.  Conditional update will  have a descriptive message and reference URL which links to more thorough explanation for the recommendation removal

Reviewer: nit: the proposed text is talking about a singular risk, but a given A->B update may be exposed to several known risks, e.g. `4.10.30 -(blocked: StaleInsightsRunLevelLabel, StaleSELinuxPolicies)-> 4.11.3`.  But I expect folks will get the idea with either singular-risk or plural-risks phrasing.

ifndef::openshift-origin[]

[id="fast-stable-channel-strategies_{context}"]
== Choosing the correct channel for your cluster

> Because update recommendations available in both `fast-{product-version}` and `stable-{product-version}` are fully supported and benefit equally from extensive data analysis the choice between the two channels should be guided based on your specific needs rather than which channel is the default or qualities implied by the channel names. Largely this is a choice between updating to the latest GA versions or waiting while Red Hat collects additional data before promoting to the stable channel.

Reviewer 5: "Because update recommendations available in both `fast-{product-version}` and `stable-{product-version}` are fully supported and benefit equally from the extensive testing and qualification Red Hat performs prior to release, the choice....."

We can't say they equally benefit from data analysis because there is a differing level of analysis carried out.

Author: Yeah it's a very nuanced point. Things make it to fast before we have data and we've analyzed it, however as soon as we decide a release is bad, whether it's made it to the stable channel or not, we remove update recommendations everywhere including the fast channel.

I really struggle with how to make this point clear but I think it's valuable to explain because I believe it bolsters the notion that fast is faster but not necessarily less stable. Other ways I've thought of are along the lines of "Fast channel contains no known serious regressions, the difference with the stable channel is that not only is it free of known serious regressions, but we have collected sufficient data that if there were detectable regressions we should know about them."

Reviewer 6: We should try to set the tone throughout that `fast` is THE channel to use and that `stable` is an "add on" for those  exceptional cases where someone can wait. Unfortunately though, `stable` is the default which kind of runs counter to that strategy.

Customers can improve this process by configuring pre-production systems on the `fast-{product-version}` channel, configuring production systems on the `stable-{product-version}` channel, and participating in the Red Hat connected customer program. Red Hat uses this program to observe the impact of updates on your specific hardware and software configurations.

Additionally, there are several factors which may lead an organization to move even production clusters to the fast channel either permanently or temporarilly including:
 - The desire to apply a specific fix known to affect your environment without delay.
 - Application of CVE fixes without delay -- Because even CVE fixes may introduce regressions we still apply promotion delays to z-streams with CVE fixes.
 - Internal pre-prod promotion processes -- If it takes your organization several weeks to promote updates between environments it would be best to use the latest available rather than introducing additional weeks or months of delay.

=== Minor version initial update rollouts

> Because the scope of changes associated with a z-stream update is orders of magnitude lower than the scope of changes between any two minor versions our data collection requirements are similarly orders of magnitude different as well. While we can safely promote z-stream updates within a week or two of GA we cannot do the same with minor version updates. Therefore, it's expected that the initial availability of update recommendations to the latest minor may take 45-90 days.

Author: Yup.
